File: microservice-architecture/outbox-pattern/go/src/outbox.go
// Package src 实现了 Outbox 模式（Outbox Pattern）的核心逻辑。
//
// 【设计模式】
//   - 观察者模式：outbox 事件被 relay 扫描并发布到 broker。
//   - 命令模式：OutboxEvent 将事件封装为数据对象，relay 异步执行发布。
//
// 【架构思想】
//   解决"写库成功 + 发消息失败"导致的数据与事件不一致问题。
//
// 【开源对比】
//   - Go 生态中通常用 Kafka + outbox 表 + 定时任务实现
//   - watermill-io/watermill：Go 消息框架，支持 outbox 模式
//   本示例用内存 slice 模拟。
package src

// Order 订单实体。
type Order struct {
	OrderID string // 订单ID
	Status  string // 订单状态
}

// OutboxEvent outbox 事件记录。
// 【设计模式】命令模式：将事件封装为数据对象，relay 异步发布。
type OutboxEvent struct {
	EventID     string // 事件唯一ID
	AggregateID string // 聚合根ID（订单ID）
	EventType   string // 事件类型
	Status      string // 发布状态：pending / published
}

// OutboxService Outbox 服务。
type OutboxService struct {
	orders []Order       // 模拟 orders 表
	outbox []OutboxEvent // 模拟 outbox 表
}

// MemoryBroker 内存消息代理（模拟 Kafka / RabbitMQ）。
type MemoryBroker struct {
	published []string // 已发布的事件ID列表
}

// NewOutboxService 创建 Outbox 服务。
func NewOutboxService() *OutboxService {
	return &OutboxService{}
}

// NewMemoryBroker 创建内存消息代理。
func NewMemoryBroker() *MemoryBroker {
	return &MemoryBroker{}
}

// CreateOrder 创建订单，同时写入 orders 和 outbox（模拟同一事务）。
func (s *OutboxService) CreateOrder(orderID string) {
	// 写入订单
	s.orders = append(s.orders, Order{OrderID: orderID, Status: "CREATED"})
	// 写入 outbox 事件（同一"事务"）
	s.outbox = append(s.outbox, OutboxEvent{
		EventID:     "EVT-" + orderID,
		AggregateID: orderID,
		EventType:   "order_created",
		Status:      "pending",
	})
}

// RelayPending relay 中继：扫描 pending 事件，发布后标记 published。
func (s *OutboxService) RelayPending(broker *MemoryBroker) {
	for i := range s.outbox {
		if s.outbox[i].Status == "pending" {
			// 发布到消息中间件
			broker.published = append(broker.published, s.outbox[i].EventID)
			// 标记为已发布
			s.outbox[i].Status = "published"
		}
	}
}

// Orders 获取所有订单。
func (s *OutboxService) Orders() []Order { return s.orders }

// Outbox 获取所有 outbox 事件。
func (s *OutboxService) Outbox() []OutboxEvent { return s.outbox }

// Published 获取已发布的事件ID列表。
func (b *MemoryBroker) Published() []string { return b.published }

File: microservice-architecture/outbox-pattern/c/src/outbox.c
/**
 * @file outbox.c - Outbox 模式（Outbox Pattern）的 C 语言实现
 *
 * 【设计模式】
 *   - 命令模式：OutboxEvent 将事件封装为结构体。
 *   - 观察者模式：relay 扫描 outbox 并发布到 broker。
 *
 * 【架构思想】
 *   C 语言用数组模拟数据库表，strcpy/strcmp 操作字符串字段。
 *
 * 【开源对比】
 *   实际工程中 outbox 表存储在 MySQL/PostgreSQL，
 *   relay 通过 SELECT ... WHERE status='pending' FOR UPDATE 扫描。
 */

#include "func.h"

#include <string.h>

/** 初始化 outbox 服务 */
void outbox_service_init(OutboxService *service)
{
    service->order_count = 0;
    service->outbox_count = 0;
}

/** 初始化消息代理 */
void broker_init(MemoryBroker *broker)
{
    broker->count = 0;
}

/**
 * 创建订单。同时写入 orders 和 outbox（模拟同一数据库事务）。
 */
void outbox_create_order(OutboxService *service, const char *order_id)
{
    /* 写入订单 */
    strcpy(service->orders[service->order_count].order_id, order_id);
    strcpy(service->orders[service->order_count].status, "CREATED");
    service->order_count++;

    /* 写入 outbox 事件（同一"事务"） */
    strcpy(service->outbox[service->outbox_count].event_id, "EVT-");
    strcat(service->outbox[service->outbox_count].event_id, order_id);
    strcpy(service->outbox[service->outbox_count].aggregate_id, order_id);
    strcpy(service->outbox[service->outbox_count].event_type, "order_created");
    strcpy(service->outbox[service->outbox_count].status, "pending");
    service->outbox_count++;
}

/**
 * relay 中继：扫描 pending 事件，发布到 broker 后标记 published。
 * 已标记 published 的事件不会被重复发布（安全重跑）。
 */
void outbox_relay_pending(OutboxService *service, MemoryBroker *broker)
{
    for (int i = 0; i < service->outbox_count; i++) {
        if (strcmp(service->outbox[i].status, "pending") == 0) {
            /* 发布到消息中间件 */
            strcpy(broker->published[broker->count++], service->outbox[i].event_id);
            /* 标记为已发布 */
            strcpy(service->outbox[i].status, "published");
        }
    }
}
